lib/txgh/diff_calculator.rb

```diff
@@ -0,0 +1,100 @@
+module Txgh
+  class DiffCalculator
+    class << self
+      def compare(head_phrases, diff_point_phrases)
+        new(head_phrases, diff_point_phrases).compare
+      end
+    end
+
+    INCLUDED_STATES = [:added, :modified]
+
+    attr_reader :head_phrases, :diff_point_phrases
+
+    def initialize(head_phrases, diff_point_phrases)
+      @head_phrases = head_phrases
+      @diff_point_phrases = diff_point_phrases
+    end
+
+    def compare
+      join_diffs(compare_head, compare_diff_point)
+    end
+
+    private
+
+    def join_diffs(diff1, diff2)
+      INCLUDED_STATES.each_with_object({}) do |state, ret|
+        ret[state] = diff1[state] + diff2[state]
+      end
+    end
+
+    def compare_head
+      Hash.new { |hash, key| hash[key] = [] }.tap do |diff|
+        head_to_diff_point.each do |phrase, state, old_phrase|
+          diff[state] << phrase
+        end
+      end
+    end
+
+    def compare_diff_point
+      Hash.new { |hash, key| hash[key] = [] }.tap do |diff|
+        diff_point_to_head.each do |phrase, state, old_phrase|
+          diff[state] << phrase
+        end
+      end
+    end
```

Comment:
`compare_diff_point` and `compare_head` are functionally identical; consider abstracting and supplying the enumeration receiver as an argument